### ISSUE TITLE
Change views on ViewPort device

### ISSUE THREAD

**Issue by AnthonyBorton** [briis/unifiprotect#178]
Firstly, thanks to all involved in contributing to this project. 🙇 

I have a feature request you might wish to consider.

I have a UDM Pro and use a [Unifi Protect ViewPort](https://store.ui.com/collections/unifi-protect-accessories/products/unifi-protect-viewport) to display the camera feeds on a screen using HDMI. I would love to be able to switch the view that is being shown on the viewport using automation/scripts in HA.  
Possible scenarios:

- During the day I would want to see one arrangement of camera feeds and then after hours, I would like a different view.
- Based on motion detection, I would like to change the view on the viewport to have that camera displayed prominently.

I'm not sure if this functionality is accessible programmatically but wanted to put this out there for consideration.

**Comment by briis** [comment]
Hi Anthony,
Thanks and thanks for the suggestion. I don't have access to the Viewport device and as such I have never played with it.
I understand what it does, and I assume the arrangement of Cameras are set in the Protect APP (or Web). 

Can you define more than one arrangement of cameras in the APP?
If yes, how do you switch between them now?

And highlighting the Camera with motion is not a feature that is built-in today?

I am asking this, as I would assume what you are asking for should already have been done by Ubiquiti, but I guess it is not, since you ask. So I am not convinced that Home Assistant is the right way to go to fix this, but I might be wrong.

With that said, I think I remember seeing some references to Viewport in the API data, so it might be possible to do what you want, but I am not sure this is the right place to do it. And I would need to purchase a Viewport device first.

**Comment by AnthonyBorton** [comment]
Thanks for the fast reply @briis.  Let me provide a little more information and clarity around my scenario.

When you're looking at the Protect Web UI, you can add multiple views and configure different layouts for your cameras.
![image](https://user-images.githubusercontent.com/20076700/104406373-e0d83880-55aa-11eb-920f-6d489692282d.png)

For example, in this layout, one camera is given proportionally more screen space then the other.
![image](https://user-images.githubusercontent.com/20076700/104406549-44626600-55ab-11eb-86dd-33cb629d7c5b.png)
Based on something eg. motion or time of day, I would like to change the view from "view A" which gives camera 1 dominance, to "view B" which has a different camera filling most of the screen.

When you're using the Protect WebUI, you can click the dropdown at the top to switch between layouts in the WebUI.
![image](https://user-images.githubusercontent.com/20076700/104406619-74116e00-55ab-11eb-80f9-5092ed8a1d59.png)

Now in terms of the ViewPort device, the way you choose what layout it should be showing is by using the Web UI and doing the following;
1. Choose the layout you want
2. Click the icon highlighted below
3. Click the Viewport device from the list.
![image](https://user-images.githubusercontent.com/20076700/104406754-bf2b8100-55ab-11eb-8dd9-79b3ffcaab0f.png)

I assume this should then just pass the view name (or ViewID) to the specific viewport device. 

Ideally, what I would love to see your HA integration be able to do is allow me to do something like the following
```unifiprotect.setviewportview(viewportid, viewname)```

I don't know if others might find this feature helpful but if it's not too much effort, I think it would be great.

**Comment by briis** [comment]
OK, understand what you want to do.
I will see if I can get my hands on a Viewport device, and then find what is possible to do.
I can't give you a timeframe for this right now.

**Comment by jonbloom** [comment]
@briis I just picked up a VP, happy to help however I can.

**Comment by briis** [comment]
I will need to write a little program that can dump the raw json file from the API, so that I can see how viewports are stored, thus making it possible to change them. Stay tuned...

**Comment by jonbloom** [comment]
Figured it out. Send a `PATCH` to `/api/viewer/<viewport_id>` with the following body:

`{liveview: <liveview_id>}`

The liveview IDs and ViewPort IDs are present in the bootstrap API response under the `liveviews` and `viewers` keys respectively. 

I see this working as a `media_player` in HA, where you can set the "source" to any of the liveviews in the system.

**Comment by briis** [comment]
As I don't have a VP, could you dump what you see in the `viewers` key? (Mine is just an empty array.

> I see this working as a media_player in HA, where you can set the "source" to any of the liveviews in the system.

That is interresting, have your tried to do it? What API call would you make to stream a liveview? (instead of the RTSP stream)

**Comment by jonbloom** [comment]
Here's the `viewers` key:
```json
[
  {
    "mac": "0123456789AB",
    "host": "10.0.100.137",
    "connectionHost": "10.0.100.5",
    "type": "UFP Viewport",
    "name": "Viewport",
    "upSince": 1621285777192,
    "lastSeen": 1621339782192,
    "connectedSince": 1621285848053,
    "state": "CONNECTED",
    "hardwareRevision": null,
    "firmwareVersion": "1.2.38",
    "firmwareBuild": "fc5e3ecd.200831.505",
    "isUpdating": false,
    "isAdopting": false,
    "isAdopted": true,
    "isAdoptedByOther": false,
    "isProvisioned": false,
    "isRebooting": false,
    "isSshEnabled": true,
    "canAdopt": false,
    "isAttemptingToConnect": false,
    "streamLimit": 16,
    "softwareVersion": "1.0.0-viewport",
    "wiredConnectionState": {
      "phyRate": 1000
    },
    "liveview": "5fb1a2dd0013e003e7000569",
    "id": "6090744802439003e7011c1e",
    "isConnected": true,
    "modelKey": "viewer"
  }
]
```

By "any of the `liveview`s in the system", I was referring to those in Protect. Similar to how my Roku TV can have it's "channel" (source) set via HA, it would be nice to select the live view from Protect per ViewPort.

**Comment by briis** [comment]
Thanks for this.

Ahh for a moment there I thought you had figured out how to use the liveview directly as a video source in Home Assistant 😁
Anyway, it should be doable, the issue I see is how to tell what liveviews you have defined, and then present those to you, when you want to switch.

One way could be to define a switch for each view defined, another way could be to create a sensor with the current liveview and then show the rest as attributes of that senor and then use a service to set the view. Or do you have any good ideas?

**Comment by jonbloom** [comment]
I was thinking to use the `media_player` sources_list. See [here](https://developers.home-assistant.io/docs/core/entity/media-player#properties) and [here](https://developers.home-assistant.io/docs/core/entity/media-player#select-source). 

Workflow:
1. On integration load, enumerate `liveviews` and build mapping of name -> id
2. Create (or update) `media_player` entity for each ViewPort detected
3. Set list of sources for each `media_player` entity to list of `liveviews`
4. On source change, retrieve liveview `id` from mapping, send PATCH to API

Happy to discuss in more real-time, hit me up on Discord if you'd like.

**Comment by briis** [comment]
Thanks Jon,
Let me work a bit on this, and then I will back or hit you on Discord.

**Comment by jonbloom** [comment]
@briis Hi! Any updates to share on this?

**Comment by briis** [comment]
I have been busy with work, so not a lot of time to work on this. I have just ordered a VP, so once I get this, I will continue the work.

**Comment by briis** [comment]
I have just released a 0.9.0 Beta 1, which adds initial support for the UFP Viewport. If you use HACS, just make sure you have enabled to get Beta Releases, and it should show up.

When you select a new View in the Source List of the Media Player, there is a delay for up to 30 sec. before the title and the list is updated. I would like it to be faster, and I look in to that, but for now it is working.

Please give me your feedback on how this works. Thanks.

Below is an abstract of the Changelog.

* `NEW`: This release adds support for the UFP Viewport device. This is done by adding the `media_player` platform, from where the views defined in Unifi Protect can be selected as source. When selecting the source, the Viewport will change it's current view to the selected source. The `media_player` platform will only be setup if UFP Viewports are found in Unfi Protect.
* `NEW`: As part of the support for the UFP Viewport, there also a new service being created, called `unifiprotect.set_viewport_view`. This service requires two parameters: The `entity_id` of the Viewport and the `view_id` of the View you want to set. `view_id` is a long string, but you can find the id number when looking at the Attributes for the media_player.

**Comment by briis** [comment]
A couple of observations from my side during testing:
- Even though I can get and set the current Liveview, the UI in the Protect APP is not updated, which indicates that this device works a bit different than all the other devices. It might have its own API somewhere, but I have not found that yet. (Maybe a WireShark expert can figure something out)
- The delay in updating the HA UI, is most likely caused by the above.

**Comment by briis** [comment]
I have changed my mind. The media player was just used as an entity type, because this was the only one who could provide a list of values to select from. But with Home Assistant 2021.7 out tomorrow, there is now a new entity type called `select`. It is actually the same as an `input_select`, with the difference being that it can be programmed from an Integration. I believe this is the right approach, because this is basically just what we want: A possibility to select a named View, and send that to the Viewport. There will be a select entity created for each Viewport found in Protect.

Going forward I will add more select entities, to replace some of the switches, where I need to create a switch for each possible value (Recording Mode). But that will be future.

If you have not installed Beta 1, then wait until Beta 2, then you don't have to remove the Media Player.

**Comment by jonbloom** [comment]
This is awesome! I've found that switching the VP feed from HA is pretty much instant, but like you said, HA updating from changing the feed in Protect is not as quick. I haven't done much digging, does the integration use the websocket feed for events? There likely would be an event to listen for for VP feed changes. Views are updating on the Protect side after a refresh, which is 100% ok with me.

My main use case was to switch my VP to a full screen view of my front lawn when the G4-Pro detects a person in it, then switch back some time later.

+1 on the `select`, this is a pretty great use case of that new entity domain.

**Comment by briis** [comment]
Glad you like it, and also support the `select` entity move. I have done most of the work on that, and it works as speedy as the Media Player implementation. Hopefully I can get the Beta 2 out later today.

There is something different with the Viewport as compared to a Camera. If you set, lets say a Recordning Mode, from HA, then that is instantly also changed in the APP, but that is not the case for the Viewport. And I do believe the difference is, that there websocket event going on. I monitored the log for a while, and no matter where you change the View from, there is no entry in the event log. So we might just have to live with that for now.

**Comment by briis** [comment]
Will be released in 0.9.0, which comes out within the next week, so I will close this now.
